### PR TITLE
fix(threads/cortex_m): don't save high regs on first schedule

### PR DESCRIPTION
**Pull request by elenaf9**
# Description

I noticed that threading on the nrf5340 is currently broken/ hard faults.

Some debugging showed that the problem is that on run, the `PendSV` tries to stored the r4-r11 registers in a null pointer.
This didn't cause any problems on the rpi-pico or nrf54840, but apparently for the nrf5340/ CortexM33  it's problematic.

To fix this, we have to skip storing/ loading of r4-r11 if the pointer is zero, i.e. when there was no previously running thread whose context has to be saved.

An alternative would be to directly in the `sched` code when `current_pid.is_none()` return 0 so that no context save/ reload happens, but then the `psp` update would also be skipped and we'd manually have to do that before returning the `0`.

The PR also includes some small optimizations/ clean-ups in `sched`.

## Issues/PRs references

The bug was a side-effekt of #262.

## Change checklist

- [x] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](../book/src/coding-conventions.md).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by kaspar030** [comment]
This looks lije a special case for the first schedule, which hits only once. How does the cycle count look for this? If it is measurably slower, I'd rather have some more logic in `start_threading()`.

**Comment by kaspar030** [comment]
But nice catch! :)

**Comment by elenaf9** [comment]
> This looks lije a special case for the first schedule, which hits only once. How does the cycle count look for this? If it is measurably slower, I'd rather have some more logic in `start_threading()`.

Not sure if it's possible to solve it in `start_threading`. It could be solved it in the else branch inside `sched` when we check if `current_pid` is None: https://github.com/future-proof-iot/RIOT-rs/commit/e61ea5db21b40d24dff5d2e6a0548a1010630f74. But this impacts compiler optimizations.

Depending on which board (so far only rpi-pico and nrf53840dk; don't have an nrf52840dk with me right now) I test, one or the other solution performs better. But both variations overall don't negatively impact performance, because the PR includes some other small optimizations within `sched`. On rpi-pico `bench_sched_yield` 293 -> 291 ticks per context switch.

File: src/riot-rs-threads/src/arch/cortex_m.rs [hunk 1]
```diff
@@ -183,25 +192,22 @@ unsafe fn sched() -> u128 {
                 }
             };
 
-            let current_high_regs;
+            let mut current_high_regs = core::ptr::null();
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/arch/cortex_m.rs:199]: `PendSV()` checks if the returned `r1` is zero, and skips saving and restoring the context if that's the case. I added a comment. Not sure what you mean with relevant docs?

[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/arch/cortex_m.rs:208]: Wdyt about d79fff9? :)

**Comment by kaspar030** [comment]
LGTM, please squash!

**Pull request merged**